plugins/modem-manager/fu-mm-device.c

```diff
@@ -0,0 +1,668 @@
+					    error);
+	if (locker == NULL)
+		return FALSE;
+	if (!fu_mm_device_at_cmd (self, "AT", error))
+		return FALSE;
+	if (!fu_mm_device_at_cmd (self, self->detach_fastboot_at, error)) {
+		g_prefix_error (error, "rebooting into fastboot not supported: ");
+		return FALSE;
+	}
+
+	/* success */
+	fu_device_set_remove_delay (device, FU_DEVICE_REMOVE_DELAY_RE_ENUMERATE);
+	fu_device_add_flag (device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	return TRUE;
+}
+
+static gboolean
+fu_mm_device_detach (FuDevice *device, GError **error)
+{
+	FuMmDevice *self = FU_MM_DEVICE (device);
+	g_autoptr(FuDeviceLocker) locker  = NULL;
+
+	locker = fu_device_locker_new (device, error);
+	if (locker == NULL)
+		return FALSE;
+
+	/* This plugin supports currently two methods to download firmware:
+	 * fastboot and qmi-pdc. A modem may require one of those, or both,
+	 * depending on the update type or the modem type.
+	 *
+	 * The first time this detach() method is executed is always for a
+	 * FuMmDevice that was created from a MM-exposed modem, which is the
+	 * moment when we're going to decide the amount of retries we need to
+	 * flash all firmware.
+	 *
+	 * If the FuMmModem is created from a MM-exposed modem and...
+	 *  a) we only support fastboot, we just trigger the fastboot detach.
+	 *  b) we only support qmi-pdc, we just exit without any detach.
+	 *  c) we support both fastboot and qmi-pdc, we will set the
+	 *     ANOTHER_WRITE_REQUIRED flag in the device and we'll trigger
+	 *     the fastboot detach.
+	 *
+	 * If the FuMmModem is created from udev events...
+	 *  d) it means we're in the extra required write that was flagged
+	 *     in an earlier detach(), and we need to perform the qmi-pdc
+	 *     update procedure at this time, so we just exit without any
+	 *     detach.
+	 */
+
+        /* FuMmDevice created from MM... */
+	if ((self->omodem != NULL)) {
+                /* both fastboot and qmi-pdc supported? another write required */
+                if ((self->update_methods & MM_MODEM_FIRMWARE_UPDATE_METHOD_FASTBOOT) &&
+                    (self->update_methods & MM_MODEM_FIRMWARE_UPDATE_METHOD_QMI_PDC)) {
+                        g_debug ("both fastboot and qmi-pdc supported, so the upgrade requires another write");
+                        fu_device_add_flag (device, FWUPD_DEVICE_FLAG_ANOTHER_WRITE_REQUIRED);
+                }
+                /* fastboot */
+                if (self->update_methods & MM_MODEM_FIRMWARE_UPDATE_METHOD_FASTBOOT)
+                        return fu_mm_device_detach_fastboot (device, error);
+                /* otherwise, assume we don't need any detach */
+                return TRUE;
+        }
+
+        /* FuMmDevice created from udev...
+         * assume we don't need any detach */
+        return TRUE;
+}
+
+typedef struct {
+	gchar  *filename;
+	GBytes *bytes;
+} FuMmFileInfo;
+
+static void
+fu_mm_file_info_free (FuMmFileInfo *file_info)
+{
+	if (file_info != NULL) {
```

Comment:
I don't think you need to check for NULL, g_autoptr macros do it for you.

Reply:
ok